tst/unit/test_sparse_pack.cpp

```diff
@@ -77,13 +77,78 @@ struct v5 : public parthenon::variable_names::base_t<false> {
   static std::string name() { return "v5"; }
 };
 
+using parthenon::variable_names::ANYDIM;
+struct v7 : public parthenon::variable_names::base_t<false, ANYDIM, 3> {
+  template <class... Ts>
+  KOKKOS_INLINE_FUNCTION v7(Ts &&...args)
+      : parthenon::variable_names::base_t<false, ANYDIM, 3>(std::forward<Ts>(args)...) {}
+  static std::string name() { return "v7"; }
+};
+
 } // namespace
 
 TEST_CASE("Test behavior of sparse packs", "[SparsePack]") {
+  constexpr int N = 6;
+  constexpr int NDIM = 3;
+  constexpr int NBLOCKS = 9;
+
+  GIVEN("A tensor variable on a mesh") {
+    const std::vector<int> tensor_shape{N, N, N, 3, 3};
+    Metadata m_tensor({Metadata::Independent}, tensor_shape);
+    auto pkg = std::make_shared<StateDescriptor>("Test package");
+    pkg->AddField<v7>(m_tensor);
+    BlockList_t block_list = MakeBlockList(pkg, NBLOCKS, N, NDIM);
+
+    MeshData<Real> mesh_data("base");
+    mesh_data.Set(block_list, nullptr);
+
+    WHEN("We initialize the independent variables by hand and deallocate one") {
+      auto ib = block_list[0]->cellbounds.GetBoundsI(IndexDomain::entire);
+      auto jb = block_list[0]->cellbounds.GetBoundsJ(IndexDomain::entire);
+      auto kb = block_list[0]->cellbounds.GetBoundsK(IndexDomain::entire);
+      for (int b = 0; b < NBLOCKS; ++b) {
+        auto &pmb = block_list[b];
+        auto &pmbd = pmb->meshblock_data.Get();
+        auto var = pmbd->Get("v7");
+        auto var5 = var.data.Get<5>();
+        int slower_rank = var5.GetDim(5);
+        int faster_rank = var5.GetDim(4);
+        par_for(
+            loop_pattern_mdrange_tag, "initializev7", DevExecSpace(), kb.s, kb.e, jb.s,
+            jb.e, ib.s, ib.e, KOKKOS_LAMBDA(int k, int j, int i) {
+              for (int l = 0; l < slower_rank; ++l) {
+                for (int m = 0; m < faster_rank; ++m) {
+                  Real n = m + 1e1 * l;
+                  var5(l, m, k, j, i) = n;
+                }
+              }
+            });
+      }
+      THEN("A sparse pack can correctly index into tensor types") {
+        auto desc = parthenon::MakePackDescriptor<v7>(pkg.get());
+        auto sparse_pack = desc.GetPack(&mesh_data);
+        int nwrong = 0;
+        par_reduce(
+            loop_pattern_mdrange_tag, "check vector", DevExecSpace(), 0,
+            sparse_pack.GetNBlocks() - 1, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+            KOKKOS_LAMBDA(int b, int k, int j, int i, int &ltot) {
+              // 0-th is ANYDIM, 1st is 3.
+              for (int l = 0; l < v7::GetDim<1>(); ++l) {
+                for (int m = 0; m < v7::GetDim<1>(); ++m) {
```

Comment:
Are these correct?
If yes, then some comment explaining why using `GetDim<1>` twice works would be good.
If this this a typo, I also suggest to update `     const std::vector<int> tensor_shape{N, N, N, 3, 3};` above to use different dimensions in the tensor indices.

Reply:
Hi, nice to meet you and  thank you for the comments! I think, it is correct because both "l" and "m" should go from 0 to 3 and 3 is v7::GetDim<1>(), please correct me if I am wrong. If you agree, I will add a comment to clarify.

Reply:
This is for the tensor variable type not the array. You could use the dimensions of the `std::vector` as @pgrete suggests, and I think that might be more clear.

The reason it's `1` both times here is that `0` is `ANYDIM`. That's supposed to test the functionality that the slowest moving index of the variable shape allowed to float---i.e., be set-able at runtime rather than compile time. (It turns out that for indexing arithmetic that size doesn't matter... same as how multi-dimensional C-style arrays can be passed into a function with one floating size.)

@mari2895 I suggest you switch to using `tensor_shape[4]` for `l` and `tensor_shape[3]` for `m` for clarity, as @pgrete suggests.

Reply:
Done

Reply:
@Yurlungur this change causes some checks to fail.

Reply:
Let's discuss over slack.

Reply:
OH it's inside of a kokkos loop!

To make them pass, assign a variable above the `par_for` like `int nl = tensor_shape[4]` and `int nm = tensor_shape[3]` and then use `nl` and `nm` in the loop itself

Reply:
Just to be 100% sure (tbh I'm still confused), the picking the correct indices in the original approach should have read

1. (using c-style indices based on `v7(l,m)`)
```
              for (int l = 0; l < v7::GetDim<0>(); ++l) {
                for (int m = 0; m < v7::GetDim<1>(); ++m) {
```
2. or (using c-style indices based on `v7(l,m)` with the `ANYDIM` offset)
```
              for (int l = 0; l < v7::GetDim<1>(); ++l) {
                for (int m = 0; m < v7::GetDim<2>(); ++m) {
```
3. or (using kji based indices in the context of our "dimensions", i.e., GetDim(1) is related to the rightmost `i` index)
```
              for (int l = 0; l < v7::GetDim<1>(); ++l) {
                for (int m = 0; m < v7::GetDim<0>(); ++m) {
```
4. or (using kji based indices in the context of our "dimensions", i.e., GetDim(1) is related to the rightmost `i` index with `ANYDIM` offset)
```
              for (int l = 0; l < v7::GetDim<2>(); ++l) {
                for (int m = 0; m < v7::GetDim<1>(); ++m) {
```
?

Based on your comment above, I think it's either option 2 or 4 but between those, I'm not sure.

Reply:
It would have been option 1 actually... Coming from this line:
```C++
  template <int idx>
  static constexpr auto GetDim() {
    return std::get<idx>(std::make_tuple(NCOMP...));
  }
```
That said, that's pretty counter to our usual way of doing things. @mari2895 I would suggest replacing this with
```C++
  template <int idx>
  static constexpr auto GetDim() {
    return std::get<sizeof...(NCOMP) - idx>(std::make_tuple(NCOMP...));
  }
```
which should match the more proper parthenon style with option 3.